docs/architecture/adr-047-handling-evidence-from-light-client.md

```diff
@@ -5,210 +5,186 @@
+```
 
-* The validator set of the common header must have at least 1/3 validator power that signed in the divergedHeaders commit
+The function takes the last header it received, the target header and compares it against all the witnesses
+it has through the following function:
 
-* One of the deterministically derived hashes (`ValidatorsHash`, `NextValidatorsHash`, `ConsensusHash`,
-`AppHash`, or `LastResultsHash`) of the header must not match:
+```golang
+func (c *Client) compareNewHeaderWithWitness(errc chan error, h *types.SignedHeader,
+	witness provider.Provider, witnessIndex int)
+```
 
-* We then take every validator that voted for the invalid header and was a validator in the common headers validator set and create `LunaticValidatorEvidence`
+The err channel is used to send back all the outcomes so that they can be processed in parallel. Invalid headers result in dropping
+the witness, lack of response or not having the headers is ignored just as headers that have the same hash. Headers, however,
+of a different hash then trigger the detection process between the primary and that particular witness.
 
-If this fails then we examine the case of Equivocation (either duplicate vote or amnesia):
+This begins with verification of the witness's header via skipping verification which is run in tandem with locating the Light Bifurcation Point
 
-*This only requires the trustedHeader and the divergedHeader*
+![](../imgs/light-client-detector.png)
 
-* if `trustedHeader.Round == divergedHeader.Round`, and a validator signed for the block in both headers then DuplicateVoteEvidence can be immediately formed
+This is done with:
 
-* if `trustedHeader.Round != divergedHeader.Round` then we form PotentialAmnesiaEvidence as some validators in this set have behaved maliciously and protocol in ADR 56 needs to be run. 
+```golang
+func (c *Client) examineConflictingHeaderAgainstTrace(
+	trace []*types.LightBlock,
+	divergentHeader *types.SignedHeader,
+	source provider.Provider, now time.Time) ([]*types.LightBlock, *types.LightBlock, error)
+```
 
-*The node does not check that there is a 1/3 overlap between headers as this may not be point of the fork and validator sets may have since changed*
+which performs the following
 
-If no evidence can be formed from a light trace, it is not a legitimate trace and thus the 
-connection with the peer should be stopped
+1. Checking that the trusted header is the same. Currently, they should not theoretically be different
+because witnesses cannot be added and removed after the client is initialized. But we do this any way
+as a sanity check. If this fails we have to drop the witness.
 
-### F1. Equivocation
+2. Querying and verifying the witness's headers using bisection at the same heights of all the
+intermediary headers of the primary (In the above example this is A, B, C, D, F, H). If bisection fails or the witness stops responding then
```

Comment:
Isn't it possible that the primary is part of the forked chain and the witness can't give us a valid bisection trace at the heights we got in the trace from the primary, even though the witness is correct?

Reply:
@ebuchman You might want to take a look at the spec (https://github.com/informalsystems/tendermint-rs/blob/master/docs/spec/lightclient/attacks/evidence-handling.md) before trying to understand this ADR.